udaykalra/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -13,11 +13,180 @@
+
+    // Obtain data from the request.
+    Collection<String> attendees = request.getAttendees();
+    long requestDuration = request.getDuration();
+
+    Collection<TimeRange> busyRanges = new HashSet<>();
+
+    // Detect relevant busy time ranges based on attendees.
+    for (Event event : events) {
+      if (!Collections.disjoint(attendees, event.getAttendees())) {
+        busyRanges.add(event.getWhen());
+      }
+    }
+
+    /* Iteratively cut availableRange's times based on
+     * overlap with busy events and store results in availableRanges.
+     */
+    for (TimeRange busyRange : busyRanges) {
+      // For a single busy event, check all time ranges for overlap and cut
+      removeOverlapsWithTimeRange(busyRange, availableRanges);
+    }
+
+    // Eliminate invalid time intervals by duration.
+    for (TimeRange toEvaluate : availableRanges) {
+      if (toEvaluate.duration() < requestDuration) {
+        availableRanges.remove(toEvaluate);
+      }
+    }
+
+    // Convert to list and sort.
+    List outList = new ArrayList(availableRanges);
+    Collections.sort(outList, TimeRange.ORDER_BY_START);
+    return outList;
+  }
+
+  /**
+   * Mutates timeRanges by removing the time range defined in possiblyOverlappingTimeRange from any
+   * overlapping TimeRange objects it contains.
+   */
+  private void removeOverlapsWithTimeRange(
+      TimeRange possiblyOverlappingTimeRange, Collection timeRanges) {
+
+    // Make a snapshot of TimeRanges for each busy check
+    Collection<TimeRange> rangesSnapshot = new HashSet<>();
+    rangesSnapshot.addAll(timeRanges);
+
+    for (TimeRange currentTimeRange : rangesSnapshot) {
+
+      // Check for overlap and adjust currentTimeRange
+      if ((currentTimeRange.overlaps(possiblyOverlappingTimeRange)) || (possiblyOverlappingTimeRange.overlaps(currentTimeRange))) {
+        timeRanges.remove(currentTimeRange);
+        replaceOverlappedTimeRange(possiblyOverlappingTimeRange, currentTimeRange, timeRanges);
+      }
+    }
+  }
+
+  /**
+   * Replaces a TimeRange in a collection to remove an overlap with overlappingTimeRange. The
+   * replacement omits the overlap through either:
+   *  1. trimming the TimeRange,
+   *  2. splitting the TimeRange and trimming the resulting TimeRanges.
+   * If events do not overlap, no changes will occur.
+   */
+  private void replaceOverlappedTimeRange(
+      TimeRange overlappingTimeRange, TimeRange toReplace, Collection<TimeRange> whereReplace) {
+    // Obtain timeRange and overlappingTimeRange start/end data..
+    int toReplaceStartMinute = toReplace.start();
+    int toReplaceEndMinute = toReplace.end();
+
+    int overlappingStartMinute = overlappingTimeRange.start();
+    int overlappingEndMinute = overlappingTimeRange.end();
+
+    int overlapType = classifyOverlap(overlappingTimeRange, toReplace);
+
+    // Case 1&4: Trim front of toReplace.
+    if ((overlapType == OVERLAP_WITH_SAME_START) || (overlapType == OVERLAPS_START)) {
```

Comment:
Don't need parenthesis around the "a == b" checks. "==" has higher precedence than "||" in Java (and in C, C++, JavaScript, Python (except || is "or"), pretty much any language you're likely to use).